Data/ID/ID.cs

```diff
@@ -0,0 +1,93 @@
+
+            /// <summary>
+            /// Creates a new instance.
+            /// </summary>
+            internal IDLimitWarningEventArgs()
+            {
+                IsHandled = false;
+            }
+
+            /// <summary>
+            /// Call if the limit has been handled and no corrective action is required by the application.
+            /// The event will continue to be raised to handlers for informational purposes (Logging, analytics, etc..)
+            /// </summary>
+            /// <remarks>
+            /// Calling <see cref="Handle"/> multiple times on a single event will raise a warning. While handling the
+            /// limit multiple times probably isn't harmful, it's not recommended.
+            /// </remarks>
+            public void Handle()
+            {
+                if (IsHandled)
+                {
+                    Log.GetLogger(this).Warning($"ID supply limit has already been handled by another listener and shouldn't get handled again.");
+                }
+
+                IsHandled = true;
+            }
+        }
+
+        /// <summary>
+        /// Triggers the first time that any instance of <see cref="AbstractIDProvider"/> passes its
+        /// SupplyWarningThreshold for the first time.
+        /// This gives the application the opportunity to react before IDs are exhausted.
+        /// </summary>
+        public static event EventHandler<IDLimitWarningEventArgs> OnIDLimitGlobalWarning;
+
+        private static readonly HashSet<AbstractIDProvider> ID_PROVIDERS = new HashSet<AbstractIDProvider>();
+
+        /// <summary>
+        /// Removes all reference to any <see cref="AbstractIDProvider"/>s that were created and resets
+        /// the <see cref="OnIDLimitGlobalWarning"/> event.
+        /// </summary>
+        /// <remarks>
+        /// Useful for when the application is soft-reloaded from inside.
+        /// </remarks>
+        public static void Dispose()
+        {
+            foreach (AbstractIDProvider provider in ID_PROVIDERS)
+            {
+                provider.OnIDLimitWarning -= Provider_OnIDLimitWarning;
+            }
+            ID_PROVIDERS.Clear();
+            OnIDLimitGlobalWarning = null;
+        }
+
+        internal static void RegisterIDProvider(AbstractIDProvider provider)
+        {
+            Debug.Assert(!ID_PROVIDERS.Contains(provider), $"{provider} is being registered with {nameof(ID)} but it already is registered!");
+            ID_PROVIDERS.Add(provider);
+            provider.OnIDLimitWarning += Provider_OnIDLimitWarning;
+        }
+
+        internal static void UnregisterIDProvider(AbstractIDProvider provider)
+        {
+            Debug.Assert(ID_PROVIDERS.Contains(provider), $"{provider} is being unregistered from {nameof(ID)} but it wasn't registered!");
+            ID_PROVIDERS.Remove(provider);
+            provider.OnIDLimitWarning -= Provider_OnIDLimitWarning;
+        }
```

Comment:
This isn't going to behave the way we want it to. We want the instance limit listeners to have an opportunity to handle the limit reached before the global event kicks off.

That behaviour allows local providers to potentially recover from supply exhaustion before the application is required to handle it at the top level.

Suggest we just create an `internal` notify global method and then we can just get rid of the need to maintain a list of provider instances.

Reply:
Hmmn... good point. Fixed